src/parquet/file/metadata.h

```diff
@@ -0,0 +1,211 @@
+  std::unique_ptr<InternalFileMetaDataReader> impl;
+};
+
+class PARQUET_EXPORT SchemaMetaDataWriter {
+ public:
+  // API convenience to get a MetaData reader
+  static std::unique_ptr<SchemaMetaDataWriter> GetSchemaMetaDataWriter();
+
+  SchemaMetaDataWriter();
+  ~SchemaMetaDataWriter();
+  const uint8_t* get_buffer();
+  size_t get_buffer_size();
+
+  // SchemaElement metadata
+  void PhysicalType(Type::type ptype);
+  void type_length(int32_t length);
+  void num_children(int32_t children);
+  void scale(int32_t scale);
+  void precision(int32_t precision);
+  void field_id(int32_t id);
+  void name(const std::string& name);
+  void ConvertedType(LogicalType::type logical_type);
+  void RepetitionType(Repetition::type repetition_type);
+
+ private:
+  // PIMPL Idiom
+  class InternalSchemaMetaDataWriter;
+  std::unique_ptr<InternalSchemaMetaDataWriter> impl;
+};
+
+class PARQUET_EXPORT RowGroupMetaDataWriter {
+ public:
+  // API convenience to get a MetaData reader
+  static std::unique_ptr<RowGroupMetaDataWriter> GetRowGroupMetaDataWriter(
+      const WriterProperties& props);
+
+  explicit RowGroupMetaDataWriter(const WriterProperties& props);
+  ~RowGroupMetaDataWriter();
+  const uint8_t* get_buffer();
+  size_t get_buffer_size();
+
+  // row-group metadata
+  void initialize_columns(int i);
+  void set_num_rows(int64_t rows);
+  void set_total_byte_size(int64_t size);
+  // column chunk
+  void set_column_file_offset(int i, int64_t offset);
+  void set_column_file_path(int i, const std::string& path);
+  // column metadata
+  void set_column_type(int i, Type::type ptype);
+  void set_column_encodings(int i, const std::shared_ptr<schema::ColumnPath>& val);
+  void set_column_path_in_schema(int i, const std::shared_ptr<schema::ColumnPath>& val);
+  void set_column_codecs(int i, const std::shared_ptr<schema::ColumnPath>& val);
+  void set_column_num_values(int i, int64_t values);
+  void set_column_total_uncompressed_size(int i, int64_t size);
+  void set_column_total_compressed_size(int i, int64_t size);
+  void set_column_data_page_offset(int i, int64_t offset);
+  void set_column_index_page_offset(int i, int64_t offset);
+  void set_column_dictionary_page_offset(int i, int64_t offset);
+  void SetColumnStatistics(int i, const ColumnStatistics& stats);
+
+ private:
+  // PIMPL Idiom
+  class InternalRowGroupMetaDataWriter;
+  std::unique_ptr<InternalRowGroupMetaDataWriter> impl;
+};
+
+class PARQUET_EXPORT FileMetaDataWriter {
+ public:
+  // API convenience to get a MetaData reader
+  static std::unique_ptr<FileMetaDataWriter> GetFileMetaDataWriter(
+      const WriterProperties& props);
+
+  explicit FileMetaDataWriter(const WriterProperties& props);
+  ~FileMetaDataWriter();
+  const uint8_t* get_buffer();
+  size_t get_buffer_size();
+
+  // file metadata
+  void set_num_rows(int64_t nrows);
+  void SetSchemas(std::vector<SchemaMetaDataWriter*>& schemas);
+  void SetRowGroups(std::vector<RowGroupMetaDataWriter*>& rowgroups);
```

Comment:
I am unsure about this from an API design perspective. To me, I would hope to see code like:

```
const RowGroupMetadataBuilder rg_builder = file_metadata_builder->AddRowGroup()
rg_builder->SetX
rg_builder->SetY
rg_builder->SetZ
rg_builder->Finish()
```